working/augmentation-libraries/feature-specification.md

```diff
@@ -622,311 +626,97 @@ It is a compile-time error if:
-    required to repeat the `late`.)*
-
-*   A getter or setter declaration is augmented by an augmenting variable.
-
-*   A late final variable with no initializer expression is augmented by an
-    augmenting variable with an initializer expression.
-    _A late final variable with no initializer has a setter, while one with an
-    initializer does not. An augmentation must not change whether there is a
-    setter._
-
-*   A `const` variable is augmented by an augmenting getter. **(TODO: Can a
-    const variable be augmented by another const variable, changing its value,
-    or is that too weird?)**
-
-*   An `abstract` variable is augmented with a non-abstract variable.
-
-*   An `external` variable is augmented with an `abstract` variable.
+*   A function is not complete after all augmentations are applied, unless it
+    is in a context where it can be abstract. *Every function declaration
+    eventually needs to have a body filled in unless it's an instance method
+    that can be abstract. In that case, if no declaration provides a body, it
+    is considered abstract.*
 
-### Augmenting enum members
-
-Some enum members can not be augmented: It is a compile-time error if an
-augmenting declaration in an enum declaration (introductory or augmenting)
-has the name `values`, `index`, `hashCode`, or `==`.
-
-*It has always been an error for an enum declaration to declare a member
-named `index`, `hashCode`, `==`, or `values`, and this rule just clarifies
-that this error is applicable for augmenting declarations as well.*
-
-Enum values can _only_ be augmented by enum values, and the implicit getter
-introduced by them is not augmentable. The only thing you are allowed to do
-when augmenting an enum value is add metadata annotations or doc comments.
+### Augmenting variables
 
-When augmenting an enum value, no constructor invocation should be provided.
-The original value is always used, and the explicit constructor invocation (if
-present) should not be copied.
+A class-like augmentation can add *new* variables (i.e. instance or static
+fields) to the type being augmented:
 
-New enum values may be defined in an augmenting enum, and they will be appended
-to the current values of the declaration in augmentation application order.
-
-Augmenting an existing enum value never changes the order in which it appears in
-`values`.
+```dart
+class C {}
 
-For example:
+augment class C {
+  int x = 3;
 
+  static int y = 4;
+}
 ```
-// main.dart
-part 'a.dart';
-part 'c.dart';
 
-enum A {
-  first,
-  second.custom(1);
+Variable declarations themselves can't be directly augmented, with the exception
+of abstract fields.
 
-  final int b;
+*A variable declaration implicitly has code for the synthesized getter and
+setter that access and modify the underlying backing storage. Since we don't
+allow augmentations to replace code, that implies that augmentations can't
+change variables. So we don't allow them to be augmented.*
```

Comment:
That makes sense, except for the lack of ability to provided an initializer expression. But then just make it an abstract variable, and have the augmentation add an actual variable and getter/settter forwarding.

And the `augment abstract ... var` is really just a way to augment with abstract setters and getters.

Reply:
> But then just make it an abstract variable, and have the augmentation add an actual variable and getter/settter forwarding.

Right. We don't *need* to support augmenting a getter/setter with a variable because you can accomplish the same thing with a little indirection. So I figured I'd keep it simple and get rid of augmenting variables completely.